androguard/core/apk/__init__.py

```diff
@@ -1501,35 +1504,256 @@ def is_androidtv(self) -> bool:
+        public_key = loaded_cert.public_key
+        key_type = loaded_cert.algorithm
+        # Verify the signature
+        try:
+            if key_type == 'rsa':
+                asymmetric.rsa_pkcs1v15_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            elif key_type == 'dsa':
+                asymmetric.dsa_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            elif key_type == 'ec':
+                asymmetric.ecdsa_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            else:
+                raise ValueError(f"Unsupported key algorithm: {key_type}")
+            matching_certificate_verified = matching_certificate.chosen.dump()
+        except errors.SignatureError:
+            logger.info(
+                f"The public key of the certificate:{hashlib.sha256(matching_certificate.chosen.dump()).hexdigest()} is not associated with the signature!")
+        return matching_certificate_verified
 
-        pkcs7obj = cms.ContentInfo.load(pkcs7message)
-        # TODO: should be returning the matching cert here!
-        # https://github.com/androguard/androguard/pull/1038
-        if len(pkcs7obj['content']['certificates']) > 1:
-            logger.warning(f"Multiple certificates found. Returning the first one!")
-        cert = pkcs7obj['content']['certificates'][0].chosen.dump()
-        return cert
+    @staticmethod
+    def get_hash_algorithm(signer_info):
+        # Determine the hash algorithm from the SignerInfo
+        digest_algorithm = signer_info['digest_algorithm']['algorithm'].native
+        # Map the digest algorithm to a hash function
+        hash_algorithms = {
+            'md5': md5,
+            'sha1': sha1,
+            'sha256': sha256,
+            'sha224': sha224,
+            'sha384': sha384,
+            'sha512': sha512
+        }
+        if digest_algorithm not in hash_algorithms:
+            raise ValueError(f"Unsupported hash algorithm: {digest_algorithm}")
+        return digest_algorithm
+
+    def find_certificate(self, signed_data_certificates, signer_info):
+        """
+        From the bag of certs, obtain the certificate referenced by the SignerInfo.
+
+        Args:
+            signed_data_certificates: List of certificates in the SignedData.
+            signer_info: SignerInfo object containing the issuer and serial number reference.
+
+        Returns:
+            The matching certificate if found, otherwise None.
+        """
+        matching_certificate = None
+        issuer_and_serial_number = signer_info['sid'].native
+        issuer = issuer_and_serial_number['issuer']
+        serial_number = issuer_and_serial_number['serial_number']
+
+        # Create a x509.Name object for the issuer in the SignerInfo
+        issuer_name = x509.Name.build(issuer)
+        issuer_str = self.canonical_name(issuer_name)
```

Comment:
Building a new `x509.Name` like this will break with my non-trivial examples. It's also not needed.

```suggestion
        issuer_and_serial_number = signer_info['sid'].chosen
        issuer_str = self.canonical_name(issuer_and_serial_number['issuer'])
        serial_number = issuer_and_serial_number['serial_number'].native
```